systemtest/apikeycmd_test.go

```diff
@@ -0,0 +1,114 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package systemtest_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/apm-server/systemtest/apmservertest"
+)
+
+func apiKeyCommand(subcommand string, args ...string) *apmservertest.ServerCmd {
+	cfg := apmservertest.DefaultConfig()
+	cfgargs, err := cfg.Args()
+	if err != nil {
+		panic(err)
+	}
+
+	var esargs []string
+	for i := 1; i < len(cfgargs); i += 2 {
+		if !strings.HasPrefix(cfgargs[i], "output.elasticsearch") {
+			continue
+		}
+		esargs = append(esargs, "-E", cfgargs[i])
+	}
+
+	userargs := args
+	args = append([]string{subcommand}, esargs...)
+	args = append(args, userargs...)
+	return apmservertest.ServerCommand("apikey", args...)
+}
+
+func TestAPIKeyCreate(t *testing.T) {
+	cmd := apiKeyCommand("create", "--name", t.Name(), "--json")
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+
+	var m map[string]interface{}
+	err = json.Unmarshal(out, &m)
+	require.NoError(t, err)
+	assert.Equal(t, t.Name(), m["name"])
+	assert.Contains(t, m, "id")
+	assert.Contains(t, m, "api_key")
+	assert.Contains(t, m, "credentials")
+}
+
+func TestAPIKeyCreateExpiration(t *testing.T) {
+	cmd := apiKeyCommand("create", "--name", t.Name(), "--json", "--expiration=1d")
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+
+	var m map[string]interface{}
+	err = json.Unmarshal(out, &m)
+	require.NoError(t, err)
+	assert.Contains(t, m, "expiration")
+}
+
+func TestAPIKeyInvalidateName(t *testing.T) {
+	for i := 0; i < 2; i++ {
+		cmd := apiKeyCommand("create", "--name", t.Name())
+		require.NoError(t, cmd.Run())
+	}
+
+	cmd := apiKeyCommand("invalidate", "--name", t.Name(), "--json")
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+
+	result := decodeJSONMap(t, bytes.NewReader(out))
+	assert.Len(t, result["invalidated_api_keys"], 2)
+	assert.Equal(t, float64(0), result["error_count"])
+}
+
+func TestAPIKeyInvalidateID(t *testing.T) {
+	cmd := apiKeyCommand("create", "--json")
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+	attrs := decodeJSONMap(t, bytes.NewReader(out))
+
+	cmd = apiKeyCommand("invalidate", "--json", "--id", attrs["id"].(string))
+	out, err = cmd.CombinedOutput()
+	require.NoError(t, err)
+	result := decodeJSONMap(t, bytes.NewReader(out))
+
+	assert.Equal(t, []interface{}{attrs["id"]}, result["invalidated_api_keys"])
+	assert.Equal(t, float64(0), result["error_count"])
+}
+
+func decodeJSONMap(t *testing.T, r io.Reader) map[string]interface{} {
+	var m map[string]interface{}
+	err := json.NewDecoder(r).Decode(&m)
+	require.NoError(t, err)
+	return m
+}
```

Comment:
API Key commands are only testing the console output, but not that the API Keys are actually created, invalidated, etc. in ES.
This has been implicitly tested in the python tests by using the `wait_until` functions.

Reply:
Good catch. I've added explicit "Authenticate" requests using the generated keys. Also added code to clean up API Keys for these tests.